mix/config_doc.txt
ENABLE_QCN 1 {0: disable, 1: enable}
USE_DYNAMIC_PFC_THRESHOLD 1 {0: disable, 1: enable}

PACKET_PAYLOAD_SIZE 1460 {packet size (byte)}

{for Floodgate}
USE_FLOODGATE 1 {0: don't use floodgate, 1: use floodgate}
SRCTOR_DST_PFC 0 {1: turn on srcToR->sender PFC}
SWTICH_ACK_MODE 1 {0: HOST_PER_PACKET_ACK(work under symmetric routing), 1: SWITCH_DST_CREDITT, 2: SWITCH_INGRESS_CREDIT}
SWITCH_WIN_M 1.5 {the window of switch}
SWITCH_ACK_TH_M 1.5 {the threshold of delay-credit}
RESET_ONLY_TOR_SWITCH_WIN 0 {only ToRs use floodgate (will cover floodgate config)}
SWTICH_BYTE_COUNTER 10240 {the byte counter theshold of sending credit}
SWTICH_CREDIT_INTERVAL 0 {the interval of sending credit}
SWITCH_ABSOLUTE_PSN 0 {to work under lossy network, we need to use PSN to detect packet loss}
SWITCH_SYN_TIMEOUT_US 0 {to work under lossy network, the timeout of sending SYN packets}

{you'd better view third.cc for detail when do experiments, 
for that a same setting may play different roles under different traffic patterns.}
FLOW_FROM_FILE 0 {0: flow generate from function(see third.cc for detail), 1: flow generate from input file}
FLOW_FILE mix/flow.txt {input file: flow to generate}
FLOW_CDF 4 {workload cdf of poisson}
FLOW_NUM 50000 {number of poisson flows}
SEED 1 {seed of random variable}
LOAD 0.8 {load of poisson flows}
INCAST_MIX 720 {number of incast flows for once incast}
INCAST_LOAD 0.5 {load of incastmix flows}
INCAST_CDF 9 {workload cdf of incast}

TOPOLOGY_FILE mix/topology.txt {input file: topoology}
FLOODGATE_CONFIG_FILE {input file: config of floodgate, see floodgate_config_doc.txt for detail}
TRACE_FILE mix/trace.txt {input file: nodes to monitor packet-level events (enqu, dequ, pfc, etc.), will be dumped to TRACE_OUTPUT_FILE}
TRACE_OUTPUT_FILE mix/mix.tr {output file: packet-level events (enqu, dequ, pfc, etc.)}
FCT_OUTPUT_FILE mix/fct.txt {output file: flow completion time of different flows}
PFC_OUTPUT_FILE mix/pfc.txt {output file: result of PFC}
REALTIME_BUFFER_BW_FILE mix/realtime {prefix of output file: realtime throughtput and buffer}
BW_INTERVAL {the interval of outputing realtime throughtput (us)}
BUFFER_INTERVAL {the interval of outputing realtime buffer (us)}

SIMULATOR_STOP_TIME 4.00 {simulation stop time (s)}

ROUTING_MODE 0 {0: Per-flow ECMP, 1: DRILL}
DRILL_LOAD_MODE 1 {(when DRILL) 0: use port's queuing length as load, 1: use port's received number of packets for a period of time as load}
QUEUE_STATISTIC_INTERVAL_US 100 {(when DRILL and DRILL_LOAD_MODE == 1) the interval of refreshing queue statistics}
SYMMETRIC_ROUTING_MODE 0 {0: non-symmetric routing, 1: control packets from receiver to sender will go symmetric path with data packets from sender to receiver}

CC_MODE 1 {Specifying different CC. 1: DCQCN, 3: HPCC, 7: TIMELY, 8: DCTCP}
ALPHA_RESUME_INTERVAL 1 {for DCQCN: the interval of update alpha}
RATE_DECREASE_INTERVAL 4 {for DCQCN: the interval of rate decrease}
CLAMP_TARGET_RATE 0 {for DCQCN: whether to reduce target rate upon consecutive rate decrease}
RP_TIMER 900  {for DCQCN: the interval of rate increase}
EWMA_GAIN 0.00390625 {for DCQCN and DCTCP: the g of EWMA}
FAST_RECOVERY_TIMES 1 {for DCQCN: number of times of increase for fast recovery}
RATE_AI 50Mb/s {Additive increase (not for DCTCP)}
RATE_HAI 100Mb/s {Hyper additive increase}
MIN_RATE 100Mb/s {Minimum rate}
DCTCP_RATE_AI 1000Mb/s {Additive increase for DCTCP}

ERROR_RATE_PER_LINK 0.0000 {Error rate of each link}
L2_CHUNK_SIZE 4000 {for DCQCN: chunk size}
L2_ACK_INTERVAL 1 {number of packets between ACK generation, 1 means per packet}
L2_BACK_TO_ZERO 0 {0: go-back-0, 1: go-back-N}

HAS_WIN 1 {0: no window, 1: has a window}
GLOBAL_T 1 {0: different server pairs use their own RTT as T, 1: use the max base RTT as the global T}
VAR_WIN 1 {0: fixed size of window (alwasy maximum), 1: variable window}
FAST_REACT 1 {0: react once per RTT, 1: react per ACK}
U_TARGET 0.95 {for HPCC: eta in paper}
MI_THRESH 5 {for HPCC: maxStage in paper}
INT_MULTI 1 {for HPCC: multiply the unit of txBytes and qLen in INT header}
MULTI_RATE 1 {for HPCC: 0: one rate for all hops, 1: one rate per hop}
SAMPLE_FEEDBACK 0 {for HPCC: 0: get INT per packet, 1: get INT once per RTT or qlen>0}

RATE_BOUND 1 {0: no rate limitor, 1: use rate limitor}

ACK_HIGH_PRIO 0 {0: ACK has same priority with data packet, 1: prioritize ACK}

LINK_DOWN 0 0 0 {a b c: take down link between b and c at time a. 0 0 0 mean no link down}

ENABLE_TRACE 1 {dump packet-level events or not}

KMAX_MAP 3 25000000000 400 50000000000 800 100000000000 1600 {a map from link bandwidth to ECN threshold kmax}
KMIN_MAP 3 25000000000 100 50000000000 200 100000000000 400 {a map from link bandwidth to ECN threshold kmin}
PMAX_MAP 3 25000000000 0.2 50000000000 0.2 100000000000 0.2 {a map from link bandwidth to ECN threshold pmax}
BUFFER_SIZE 20 {buffer size per switch}
QLEN_MON_FILE mix/qlen.txt {output file: result of qlen of each port}
QLEN_MON_START 2000000000 {start time of dumping qlen}
QLEN_MON_END 2010000000 {end time of dumping qlen}


mix/floodgate_config_doc.txt
13{the number of switches which use floodgate} 1{the number of switches which have some additional rules] 100[the number of voq of a voq group} 1{voq groups use dynamic hash}

{ids of switches which use floodgate:}
144
145
146
147
148
149
150
151
152
153
154
155
156

{describe the switches which need some additional rules}
144{switch_id} 2{the number of voq group} 16{the number of rules}
{groups' settings(the group_id starts from 0)}
100{the number of voq of the first group} 1{whether first group uses dynamic hash or not}
0{the number of voq of the second group} 0{whether second group uses dynamic hash or not}
{rules: allocate dst to a voq group}
0{dst_id} 1{group_id}
1 1
2 1
3 1
4 1
5 1
6 1
7 1
8 1
9 1
10 1
11 1
12 1
13 1
14 1
15 1


changelog
2020/02/27
add 40Gbps rate in lineRateValues and logic in Set function
int-header.h
void Set(uint64_t _time, uint64_t _bytes, uint32_t _qlen, uint64_t _rate){}

int-header.cc
lineRateValues[lineRate]


src/point-to-point/model/settings.cc
#include "ns3/settings.h"

namespace ns3{

	/**
	 * return true:
	 * L4 should solve dis-order of packet
	 * Receiver:
	 * 1. Reply to the corresponding ack directly after receiving the packet and set `m_received_psn.insert(curr_psn)`
	 * 2. The receiving end receives a packet marked with lasttag --> `received_last_psn_packet = true`
	 * 3. Every time a package is received, check if `m_received_last_psn_packet` and the previous packets are also received.
	 *    --> Send a lasttag ack to inform the sender that all packets have been received.
	 * Sender:
	 * 1. Send packets in order, `send_next += payload, m_sent_psn.insert(curr_psn)`
	 * 2. Received an ACK, ` m_sent psn.erase(curr_psn)`
	 * 3. The last PSN tag is sending, tag lasttag
	 * 4. Receiving the ack of lasttag tag indicates that the receiver has received it
	 * 5. Note that: for Go-Back-N that cannot be triggered by NACK, retransmission with timeout:
	 *    * `timeouted = true`
	 *    * If `send_nxt ≠ last_psn`, continue to send packets in sequence
	 *    * Otherwise, `m_timeout_psn = m_sent_psn', send packets which indicate by `m_timeout_psn`
	 */
	bool Settings::IsPacketLevelRouting(){
		switch (Settings::routing_mode){
		case Settings::ROUTING_DRILL:
			return true;
		default:
			return false;
		}
	}

	std::vector<double> Settings::bwList[NODESCALE];
	std::vector<double> Settings::QDelayList[NODESCALE];
	double Settings::host_bw_Bps;

	std::vector<double> Settings::ctrlList[NODESCALE];
	std::vector<double> Settings::tpList[NODESCALE];

	// l4, for timeout retransmission
	uint32_t Settings::rto_us = 5000;

	// for qp mode
	bool Settings::qp_mode = false;
	bool Settings::reset_qp_rate = false;

	uint32_t Settings::free_token = 9;
	uint32_t Settings::max_bdp;

	// for floodgate
	std::map<uint32_t, uint32_t> Settings::hostIp2IdMap;
	std::map<uint32_t, uint32_t> Settings::hostId2IpMap;
	// important settings
	bool Settings::use_floodgate = false;
	double Settings::switch_win_m = 1.5;
	uint32_t Settings::switch_ack_mode = Settings::SWITCH_DST_CREDIT;
	bool Settings::reset_only_ToR_switch_win = false;
	uint32_t Settings::switch_byte_counter = 1;
	double Settings::switch_credit_interval = 10;	// us
	double Settings::switch_ack_th_m = 0;		// active delay SwitchACK or not(supported only when SWTICH_ACCUMULATE_CREDIT); performance has no outstanding improvement
	uint64_t Settings::switch_absolute_psn = 0;
	uint32_t Settings::switch_syn_timeout_us = 1000;
	// fixed settings(validated performance)
	uint32_t Settings::reset_isolate_down_host = 1;		// always use 1: isolate upstream and downstream at ToR (performance is not bad, but more intuitive)

	// for PFC
	bool Settings::srcToR_dst_pfc = true;
	uint32_t Settings::pfc_dst_alpha = 8;
	double Settings::pfc_th_static = 1.5;		// should work with "USE_DYNAMIC_PFC_THRESHOLD 0" and "SRCTOR_DST_PFC 1"

	// for flow generation
	bool Settings::pktsmode;
	uint32_t Settings::maxx; // max flow size
	double Settings::homa_longflow_threshold;
	std::vector<std::pair<uint32_t,double> > Settings::cdfFromFile;
	int Settings::avgSizeFromFile;
	double Settings::threshold_slope;
	double Settings::threshold_bias;

	// for statistic
	uint32_t Settings::timeout_times = 0;
	uint32_t Settings::drop_packets = 0;
	std::ofstream Settings::switch_buffer_out[SWITCHSCALE];
	std::ofstream Settings::switch_bw_out[SWITCHSCALE];
	std::ofstream Settings::bw_out;
	std::ofstream Settings::rate_out;
	std::ofstream Settings::queuing_out;
	uint32_t Settings::bw_interval = 10;
	uint32_t Settings::buffer_interval = 10;
	uint32_t Settings::host_num = 144;
	uint32_t Settings::switch_num = 13;
	uint32_t Settings::host_per_rack = 16;
	uint32_t Settings::tor_num = 9;
	uint32_t Settings::core_num = 4;
	uint32_t Settings::max_port_length = 0;
	uint32_t Settings::max_port_index = 0;
	uint32_t Settings::max_voq_length = 0;
	uint32_t Settings::max_switch_voq_length = 0;
	uint32_t Settings::max_switch_length = 0;

	uint32_t Settings::packet_payload = 1460;
	uint32_t Settings::MTU = 1538;

	// for routing
	uint32_t Settings::routing_mode = 0;
	uint32_t Settings::drill_load_mode = 0;
	uint32_t Settings::symmetic_routing_mode = 0;
	uint32_t Settings::queue_statistic_interval_us = 10;

	bool Settings::nack_reaction = false;

}
